design/design-doc-e2e-testing.md

```diff
@@ -0,0 +1,809 @@
+             kind: NopResource
+             spec:
+               forProvider:
+                 conditionAfter:
+                   - conditionType: Ready
+                     conditionStatus: "False"
+                     time: 0s
+                   - conditionType: Ready
+                     conditionStatus: "True"
+                     time: 10s
+                   - conditionType: Synced
+                     conditionStatus: "False"
+                     time: 0s
+                   - conditionType: Synced
+                     conditionStatus: "True"
+                     time: 10s
+               writeConnectionSecretsToRef:
+                 namespace: crossplane-system
+                 name: nop-example-resource
+         - name: nopinstance2
+           base:
+             apiVersion: nop.crossplane.io/v1alpha1
+             kind: NopResource
+             spec:
+               forProvider:
+                 conditionAfter:
+                   - conditionType: Ready
+                     conditionStatus: "False"
+                     time: 0s
+                   - conditionType: Ready
+                     conditionStatus: "True"
+                     time: 10s
+                   - conditionType: Synced
+                     conditionStatus: "False"
+                     time: 0s
+                   - conditionType: Synced
+                     conditionStatus: "True"
+                     time: 10s
+               writeConnectionSecretsToRef:
+                 namespace: crossplane-system
+                 name: nop-example-resource
+   """
+ When claim gets deployed
+   """
+     apiVersion: nop.example.org/v1alpha1
+     kind: NopResource
+     metadata:
+       name: nop-example
+     spec:
+       coolField: example
+   """
+ Then claim becomes synchronized and ready
+ And claim composite resource becomes synchronized and ready
+ And composed managed resources become ready and synchronized
+```
+
+Step "claim gets deployed" is implemented in the following Go function:
+```go
+func claimGetsDeployed(ctx context.Context, rawYaml *godog.DocString) (context.Context, error) {
+	sc := ScenarioContextValue(ctx)
+	claim, err := ToUnstructured(rawYaml.Content)
+	if err != nil {
+		return ctx, err
+	}
+	sc.Claim = claim
+	return ctx, sc.Cluster.ApplyYamlToNamespace(sc.Namespace, rawYaml.Content)
+}
+```
+
+These functions are then bound to steps at the initialization:
+```go
+func InitializeScenario(ctx *godog.ScenarioContext) {
+	ctx.Before(func(ctx context.Context, scenario *godog.Scenario) (context.Context, error) {
+		sc := &scenarioContext{
+			Cluster: &cluster{
+				cli: "kubectl",
+			},
+			Namespace: fmt.Sprintf("test-%s", scenario.Id),
+		}
+		if err := sc.Cluster.createNamespace(sc.Namespace); err != nil {
+			return ctx, err
+		}
+		ctx = context.WithValue(ctx, scenarioContextKey, sc)
+		return ctx, nil
+	})
+	ctx.Step(`^claim becomes synchronized and ready$`, claimBecomesSynchronizedAndReady)
```

Comment:
The steps we made are an example of what can be done, I don't think any the framework examples has very optimized code. I see no reason why we couldn't implement a step like `<resource> becomes <conditions>...` . 

The expectation is that during the initial few tests we create we create a collection of steps which would be enough to write any test just by making a .feature, without the need to go into the steps and their implementation

Reply:
> How reusable would these bindings be?

As you can see already in this PR (features files), the steps are reused across multiple scenarios. Yes, the steps can be parametrized and formulated so that it matches our expectations...